apps/examples/smartfs/read_optimization/read_optimization_main.c

```diff
@@ -0,0 +1,273 @@
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <sys/stat.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <time.h>
+
+#include <tinyara/timer.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+#define SAMPLE_FILE	"/mnt/sample_file"
+#define TIMER_DEVNAME	"/dev/timer%d"
+
+/****************************************************************************
+ * Private data
+ ****************************************************************************/
+char g_buf[1025];
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: read_test
+ *
+ * Description: Perform a back-to-back read of the entire file's contents,
+ *              multiple times.
+ *
+ ****************************************************************************/
+
+static int read_test(int size)
+{
+	int fd;
+	int ret;
+	int itr = 0;
+
+	while (itr < 100) {
+		fd = open(SAMPLE_FILE, O_RDOK);
+		if (fd < 0) {
+			printf("Unable to open Sample File while testing, Errno = %d\n", errno);
+			return ERROR;
+		}
+
+		do {
+			ret = read(fd, g_buf, size);
+		} while (ret == size);
+
+		close(fd);
+		itr += 1;
+	}
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: create_test_file
+ *
+ * Description: Creates the test file with test data that we will use
+ *              to conduct the test.
+ *
+ ****************************************************************************/
+
+static int create_test_file(int f_size)
+{
+	int fd;
+	int ret;
+	memset(g_buf, '1', 1024);
+
+	fd = open(SAMPLE_FILE, O_WROK | O_CREAT);
+	if (fd < 0) {
+		printf("Unable to open Sample File for testing, Error = %d\n", errno);
+		return ERROR;
+	}
+
+	while (f_size >= 1024) {
+		ret = write(fd, g_buf, 1024);
+		if (ret != 1024) {
```

Comment:
It's guaranteed to write 1024 at a time? there is no possibility to write partial data? please check.

Reply:
In this case, a partial data write would mean failure at the FS end.
The aim of this app is just a simple demonstration.